### PR TITLE
Weak map feature

### LINKED ISSUE #1443: Add support for HTML5 WeakMap
### Expected
It should be possible to use HTML5 WeakMap. It would be nice to have typesafe wrapper around it.

### Actual
Support for this feature is currently missing

### Steps To Reproduce
grepping for weakmap in sources of Bridge yields no result

### PR DESCRIPTION
**Pull request by d-p-y**
Fixes #1443.

Changes proposed in this pull request:
- implements HTML5 WeakMap
- add clients tests (not sure if correct as not able to compile them without Bridge.Test)

I also propose to add such typesafe wrapper around WeakMap. I don't find any good place to add it (my understanding is that **Bridge.HTML5** may only contain delegates to native JS). 

```csharp
    /// <summary>
    /// Thin wrapper around <see cref="WeakMap"/>
    /// </summary>
    public class WeakDictionary<TKey,TValue> where TKey : class 
    {
        private WeakMap _impl;

        public WeakDictionary()
        {
             _impl = new WeakMap();
        }
        
        public WeakDictionary(IEnumerable<KeyValuePair<TKey,TValue>> items)
        {
             _impl = new WeakMap(
                 items.Select(x => new object[] {x.Key, x.Value}).ToArray()
             );
        }

        /// <summary>
        /// Clear all elements from Dictionary
        /// </summary>
        public void Clear()
        {
            //as WeakMap.clear() is not widely implemented it uses workaround
            _impl = new WeakMap();
        }
        public TValue this[TKey key]
        {
            get
            {
                return Get(key);
            }
            set
            {
                Set(key, value);
            }
        }

        public TValue Get(TKey key)
        {
            var result = _impl.Get(key);
            if (result == null)
            {
                // friendlier exception instead of default Bridge.InvalidCastException
                throw new KeyNotFoundException("No value for key present in WeakMap");
            }
            return (TValue)result;
        }

        public void Set(TKey key, TValue value)
        {
            _impl.Set(key, value);
        }

        public bool Delete(TKey key)
        {
            return _impl.Delete(key);
        }

        public bool ContainsKey(TKey key)
        {
            return _impl.Has(key);
        }
    }
```

### CODE REVIEW

File: Html5/WeakMap.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        {
+        }
+
+        /// <summary>
+        ///  Initialize WeakMap using items that are expected to be an array of the two-element-arrays. Each two-element-array is key and value respecively
+        /// Beware that it is not widely implemented (2016-05)
+        /// </summary>
+        public WeakMap(object[][] items)
+        {
+        }
```
Comment (kcudnik) [inline Html5/WeakMap.cs]: constructors also can be marked as extern

File: Html5/WeakMap.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        public virtual extern void Set<T>(T key, object value) where T : class;
+
+        /// <summary>
+        /// deletes element. Returns true if element was contained in the map. Returns false when it wasn't present in the map
+        /// </summary>
+        /// <param name="key"></param>
+        /// <returns></returns>
+        public virtual extern bool Delete<T>(T key) where T : class;
+
+        public virtual extern bool Has<T>(T key)  where T : class;
```
Comment (kcudnik) [inline Html5/WeakMap.cs]: Also add summary documentation headers to be consistent ?

**Comment by d-p-y** [comment]
Kamil, I've changed code according to your remarks

**Comment by d-p-y** [comment]
Hi, 
Question: are those pushed tests Tests/Batch1/Collections/Native/WeakMapTests.cs
 useful or I should not bother in future? I ask because I have yet another submission ready and I wonder if I should work on this or not.
Thanks
Dominik

**Comment by LeonidVeriga** [comment]
Hi @d-p-y!

Sorry for that delay in responding.
Yes, the tests you commited are very useful.

**Comment by LeonidVeriga** [comment]
Hi @d-p-y,

Merged your Pull Request by https://github.com/bridgedotnet/Bridge/commit/ad8bc27eb00db640b89426c2e1bd3fc6b1d0167b.

Thank you for your time and contribution.

**Comment by LeonidVeriga** [comment]
@d-p-y,
Regarding your suggestion on `WeakDictionary`, we think that is a good idea and you could create your own NuGet package with name like `<your brandname>.<your utils/extention name>`

**Pull request closed without merging**